File: Src/Engine/Runtime/UI/GUI/Tooltip.cpp


#include "Tooltip.h"

#include "RootControl.h"
#include "Style.h"
#include "WindowRootControl.h"
#include "Core/Input/Input.h"
#include "Core/Input/Mouse.h"
#include "Runtime/Graphics/GraphicWindow.h"
#include "Runtime/Render/2D/Render2D.h"
#include "Runtime/Render/2D/TextLayoutOptions.h"

namespace SE
{
	void Tooltip::UpdateWindowSize()
	{
		if (m_Window)
		{
			m_Window->SetClientSize(Size.operator->() * m_Window->GetDpiScale());
		}
	}
	
	void Tooltip::WrapPosition(Float2 locationSS, float flipOffset)
	{
		if (m_ShowTarget == nullptr || m_ShowTarget->RootWindow() == nullptr)
			return;

		// Calculate popup direction
		float dpiScale = m_ShowTarget->RootWindow()->GetDpiScale();
		Float2 dpiSize = Size.operator->() * dpiScale;
		Rectangle monitorBounds = Platform::GetMonitorBounds(locationSS);
		Float2 rightBottomMonitorBounds = monitorBounds.GetBottomRight();
		Float2 rightBottomLocationSS = locationSS + dpiSize;

		// Prioritize tooltip placement within parent window, fall back to virtual desktop
		if (rightBottomMonitorBounds.y < rightBottomLocationSS.y)
		{
			// Direction: up
			locationSS.y -= dpiSize.y + flipOffset;
		}
		if (rightBottomMonitorBounds.x < rightBottomLocationSS.x)
		{
			// Direction: left
			locationSS.x -= dpiSize.x + flipOffset * 2;
		}
	}

	Tooltip::Tooltip() : ContainerControl(0, 0, 300, 24)
	{
		Visible = false;;
		AutoFocus = false;
	}

	void Tooltip::Show(Control* target, Float2 location, Rectangle targetArea)
	{
		ENGINE_ASSERT(target != nullptr);

		// Ensure to be closed
		Hide();

		// Block showing tooltips when application is not focused
		if (!Platform::GetHasFocus())
			return;

		// Unlock and perform controls update
		UnlockChildrenRecursive();
		PerformLayout();

		// Calculate popup direction and initial location
		RootControl* parentWin = target->Root;
		if (parentWin == nullptr)
		{
			return;
		}
		float dpiScale = target->RootWindow()->GetDpiScale();
		Float2 dpiSize = Size.operator->() * dpiScale;
		Float2 locationWS = target->PointToWindow(location);
		Float2 locationSS = parentWin->PointToScreen(locationWS);
		m_ShowTarget = target;
		WrapPosition(locationSS);

		// Create window
		CreateWindowSettings desc = DefaultWindowSettings();
		desc.StartPosition = WindowStartPosition::Manual;
		desc.Position = locationSS;
		desc.Size = dpiSize;
		desc.Fullscreen = false;
		desc.HasBorder = false;
		desc.SupportsTransparency = false;
		desc.ShowInTaskbar = false;
		desc.ActivateWhenFirstShown = false;
		desc.AllowInput = false;
		desc.AllowMinimize = false;
		desc.AllowMaximize = false;
		desc.AllowDragAndDrop = false;
		desc.IsTopmost = true;
		desc.IsRegularWindow = false;
		desc.HasSizingFrame = false;
		desc.ShowAfterFirstPaint = true;
		m_Window = New<GraphicWindow>(desc);
		ENGINE_ASSERT_M(m_Window != nullptr, SE_TEXT("Failed to create tooltip window."));

		// Attach to the window and focus
		Parent = m_Window->GetGUI();
		Visible = true;;
		m_Window->Show();
		m_ShowTarget->OnTooltipShown(this);
	}
	
	void Tooltip::Hide()
	{
		if (!Visible)
			return;

		// Unlink
		SetIsLayoutLocked(true);
		Parent = nullptr;

		// Close window
		if (m_Window)
		{
			GraphicWindow* win = m_Window;
			m_Window = nullptr;
			win->Close();
		}

		// Hide
		Visible = false;;
	}
	
	void Tooltip::OnMouseEnterControl(Control* target)
	{
		m_LastTarget = target;
		m_TimeToPopupLeft = TimeToShow;
	}
	
	void Tooltip::OnMouseOverControl(Control* target, float dt)
	{
		if (!Visible && m_TimeToPopupLeft > 0.0f)
		{
			m_LastTarget = target;
			m_TimeToPopupLeft -= dt;

			if (m_TimeToPopupLeft <= 0.0f)
			{
				Float2 location;
				Rectangle area;
				if (m_LastTarget->OnShowTooltip(m_CurrentText, location, area))
				{
					Show(m_LastTarget, location, area);
				}
			}
		}
	}
	
	void Tooltip::OnMouseLeaveControl(Control* target)
	{
		if (Visible)
			Hide();
		m_LastTarget = nullptr;
	}
	
	void Tooltip::Update(float deltaTime)
	{
		Float2 mousePos = Input::Mouse->GetPosition();
		Float2 location = m_ShowTarget->PointFromScreen(mousePos);
		if (!m_ShowTarget->OnTestTooltipOverControl(location))
		{
			// Auto hide if mouse leaves control area
			Hide();
		}
		else
		{
			// Position tooltip when mouse moves
			WrapPosition(mousePos, 10);
			if (m_Window)
				m_Window->SetPosition(mousePos + Float2(15, 10));
		}

		ContainerControl::Update(deltaTime);
	}
	
	void Tooltip::Draw()
	{
		Style* style = Style::Current;

		// Background
		Render2D::FillRectangle(Rectangle(Float2::Zero, Size), Color::Lerp(style->BackgroundSelected, style->Background, 0.6f));
		Render2D::FillRectangle(Rectangle(1.1f, 1.1f, Width - 2, Height - 2), style->Background);

		// Padding for text
		Rectangle textRect = GetClientArea();
		textRect.Location.x += 5;
		textRect.Size.x -= 10;

		TextLayoutOptions layout;
		layout.Bounds = textRect;
		layout.HorizontalAlignment = TextAlignment::Center;
		layout.VerticalAlignment = TextAlignment::Center;
		layout.TextWrapping = TextWrapping::WrapWords;
		layout.Scale = 1.0f;
		layout.BaseLinesGapScale = 1.0f;

		// Tooltip text
		Render2D::RenderText(style->FontMedium, m_CurrentText, style->Foreground, layout);
	}
	
	bool Tooltip::OnShowTooltip(String text, Float2& location, Rectangle& area)
	{
		ContainerControl::OnShowTooltip(text, location, area);

		// It's better not to show tooltip for a tooltip.
		// It would be kind of tooltipness.
		return false;
	}
	
	void Tooltip::OnDestroy()
	{
		Hide();

		ContainerControl::OnDestroy();
	}
	
	void Tooltip::PerformLayoutBeforeChildren()
	{
		ContainerControl::PerformLayoutBeforeChildren();

		Float2 prevSize = Size;
		Style* style = Style::Current;

		// Calculate size of the tooltip
		Float2 size = Float2::Zero;
		if (style != nullptr && style->FontMedium && m_CurrentText.Length() > 0)
		{
			TextLayoutOptions layout = TextLayoutOptions::Default();
			layout.Bounds = Rectangle(0, 0, MaxWidth, 10000000);
			layout.HorizontalAlignment = TextAlignment::Center;
			layout.VerticalAlignment = TextAlignment::Center;
			layout.TextWrapping = TextWrapping::WrapWords;
			auto items = style->FontMedium->ProcessText(m_CurrentText, layout);
			for (int i = 0; i < items.Count(); i++)
			{
				auto item = &items[i];
				size.x = Math::Max(size.x, item->Size.x + 8.0f);
				size.y += item->Size.y;
			}
			//size.X += style.FontMedium.MeasureText(_currentText).X;
		}
		Size = size + Float2(24.0f);

		// Check if is visible size get changed
		if (Visible && prevSize != Size)
		{
			UpdateWindowSize();
		}
	}
} // SE

File: Src/Engine/Runtime/UI/GUI/Tooltip.cs
using System;

namespace SE.GUI
{
    /// <summary>
    /// A delayed, non-interactive tooltip overlay hosted by a managed root control.
    /// </summary>
    public sealed class Tooltip : ContainerControl
    {
        private const float HorizontalPadding = 6.0f;
        private const float VerticalPadding = 4.0f;
        private const float CharacterWidth = 7.0f;
        private const float LineHeight = 18.0f;

        private readonly Label m_TextLabel;
        private float m_TimeToPopupLeft;
        private Control? m_LastTarget;
        private Control? m_ShowTarget;

        public Tooltip()
        {
            AutoFocus = false;
            Enabled = false;
            IsScrollable = false;
            Visible = false;
            BackgroundColor = Style.Current.BackgroundNormal;
            m_TextLabel = new Label
            {
                AutoFocus = false,
                IsScrollable = false,
                TextWrapping = TextWrapping.WrapWords,
                TextColor = Style.Current.Foreground,
            };
            AddChild(m_TextLabel);
        }

        /// <summary>
        /// Gets or sets the pointer dwell time, in seconds, before the tooltip opens.
        /// </summary>
        public float TimeToShow { get; set; } = 0.3f;

        /// <summary>
        /// Gets or sets the maximum tooltip width in logical pixels.
        /// </summary>
        public float MaxWidth { get; set; } = 500.0f;

        /// <summary>
        /// Gets whether the tooltip is currently attached to a root and visible.
        /// </summary>
        public bool IsShowing => m_ShowTarget != null && Visible;

        /// <summary>
        /// Shows the tooltip for the supplied target using its <see cref="Control.TooltipText"/>.
        /// </summary>
        public void Show(Control target, Float2 location, Rectangle targetArea)
        {
            ArgumentNullException.ThrowIfNull(target);
            _ = targetArea;
            if (string.IsNullOrWhiteSpace(target.TooltipText) || target.Root == null)
            {
                Hide();
                return;
            }

            RootControl root = target.Root;
            m_TextLabel.Text = target.TooltipText;
            float availableWidth = MathF.Max(32.0f, MaxWidth);
            float rawWidth = target.TooltipText.Length * CharacterWidth + HorizontalPadding * 2.0f;
            float width = MathF.Min(availableWidth, MathF.Max(32.0f, rawWidth));
            int lineCount = Math.Max(1, (int)MathF.Ceiling(rawWidth / width));
            float height = lineCount * LineHeight + VerticalPadding * 2.0f;
            Float2 desiredPosition = target.PointToRoot(location) + new Float2(12.0f, 18.0f);
            float x = Math.Clamp(desiredPosition.X, 0.0f, MathF.Max(0.0f, root.Width - width));
            float y = Math.Clamp(desiredPosition.Y, 0.0f, MathF.Max(0.0f, root.Height - height));

            SetBounds(x, y, width, height);
            m_TextLabel.SetBounds(HorizontalPadding, VerticalPadding, width - HorizontalPadding * 2.0f, height - VerticalPadding * 2.0f);
            if (!ReferenceEquals(Parent, root))
                root.AddChild(this);
            Visible = true;
            m_ShowTarget = target;
        }

        /// <summary>
        /// Hides the tooltip and detaches it from its root.
        /// </summary>
        public void Hide()
        {
            m_ShowTarget = null;
            Visible = false;
            Parent?.RemoveChild(this);
        }

        /// <summary>
        /// Starts delayed tooltip tracking for the supplied control.
        /// </summary>
        public void OnMouseEnterControl(Control target)
        {
            ArgumentNullException.ThrowIfNull(target);
            if (ReferenceEquals(target, m_LastTarget))
                return;

            Hide();
            m_LastTarget = target;
            m_TimeToPopupLeft = MathF.Max(0.0f, TimeToShow);
        }

        /// <summary>
        /// Updates delayed tooltip tracking while the pointer remains over the supplied control.
        /// </summary>
        public void OnMouseOverControl(Control target, float deltaTime)
        {
            ArgumentNullException.ThrowIfNull(target);
            if (!ReferenceEquals(target, m_LastTarget))
                OnMouseEnterControl(target);
            UpdateTooltip(deltaTime);
        }

        /// <summary>
        /// Stops tracking a control after its pointer leaves.
        /// </summary>
        public void OnMouseLeaveControl(Control target)
        {
            if (!ReferenceEquals(target, m_LastTarget) && !ReferenceEquals(target, m_ShowTarget))
                return;

            m_LastTarget = null;
            m_TimeToPopupLeft = 0.0f;
            Hide();
        }

        public override void Update(float deltaTime)
        {
            UpdateTooltip(deltaTime);
            base.Update(deltaTime);
        }

        private void UpdateTooltip(float deltaTime)
        {
            if (m_ShowTarget != null)
            {
                if (!m_ShowTarget.IsMouseOver || !m_ShowTarget.VisibleInHierarchy)
                    Hide();
                return;
            }

            if (m_LastTarget == null || !m_LastTarget.IsMouseOver || string.IsNullOrWhiteSpace(m_LastTarget.TooltipText))
                return;

            m_TimeToPopupLeft -= MathF.Max(0.0f, deltaTime);
            if (m_TimeToPopupLeft > 0.0f)
                return;

            Show(m_LastTarget, m_LastTarget.PointFromRoot(m_LastTarget.Root?.MousePosition ?? Float2.Zero), m_LastTarget.Bounds);
        }
    }
}
